decomp/semantics/uds.py

```diff
@@ -960,37 +1188,199 @@ def _add_edge_annotation(self, edge, attrs):
-        raise an Exception.
+        Since only the document graphs are serialized, the sentence
+        graphs must also be provided to this method call in order
+        to properly associate them with their documents.
 
         Parameters
         ----------
-        digraph
-            digraph to search
+        document
+            a dictionary constructed by networkx.adjacency_data,
+            containing the graph for the document
+        sentence_graphs
+            a dictionary containing (possibly a superset of) the
+            sentence-level graphs for the sentences in the document
+        sentence_ids
+            a dictionary containing (possibly a superset of) the
+            UD sentence IDs for each graph
+        name
+            identifier to append to the beginning of node ids
         """
+        document_graph = UDSDocumentGraph.from_dict(document, name)
+        sent_graph_names = set(map(lambda node: node['semantics']['graph'], document['nodes']))
+        sent_graphs = {gname: sentence_graphs[gname] for gname in sent_graph_names}
+        sent_ids = {gname: sentence_ids[gname] for gname in sent_graph_names}
+        genre = name.split('-')[0]
+        timestamp = cls._get_timestamp_from_document_name(name)
+        return cls(sent_graphs, sent_ids, name, genre, timestamp, document_graph)
+
+    @staticmethod
+    def _get_timestamp_from_document_name(document_name):
+        timestamp = re.search('\d{8}_?\d{6}', document_name)
+        return timestamp[0] if timestamp else None
+
+    def add_sentence_graphs(self, sentence_graphs: Dict[str, UDSSentenceGraph], 
+                                  sentence_ids: Dict[str, str]) -> None:
+        """Add additional sentences to a document
 
-        root_candidates = [node for (node, degree) in digraph.in_degree() if degree == 0]
-        if len(root_candidates) != 1:
-            raise Exception('no unique root of digraph {}'.format(digraph.name))
-        return root_candidates[0]
+        Parameters
+        ----------
+        sentence_graphs
+            a dictionary containing the sentence-level graphs
+            for the sentences in the document
+        sentence_ids
+            a dictionary containing the UD sentence IDs for each graph
+        name
+            identifier to append to the beginning of node ids
+        """
+        for gname, graph in sentence_graphs.items():
+            sentence_graphs[gname].sentence_id = sentence_ids[gname]
+            sentence_graphs[gname].document_id = self.name
+            self.sentence_graphs[gname] = graph
+            self.sentence_ids[gname] = sentence_ids[gname]
+            for node in graph.semantics_nodes:
+                semantics = {'graph': gname, 'node': node}
+                document_node_name = node.replace('semantics', 'document')
+                self.document_graph.graph.add_node(document_node_name,
+                            domain='document', type='relation',
```

Comment:
The node type isn't `'relation'`, right? (Though the edge types might be. We should just inherit the type of the semantics node, right?

Reply:
Yeah, I wasn't thinking clearly.